dalm/training/generator_only/trainer.py

```diff
@@ -0,0 +1,301 @@
+    PreTrainedTokenizerBase,
+    TrainingArguments,
+)
+from trl import SFTTrainer  # type: ignore[import]
+
+logger = logging.getLogger(__name__)
+
+
+def create_datasets(
+    dataset_name: str,
+    split: str,
+    validation_split: Optional[float],
+    size_valid_set: Optional[int],
+    streaming: bool,
+    shuffle_buffer: Optional[int],
+    num_workers: int,
+    local_dataset: bool = False,
+) -> Tuple[Dataset, Dataset]:
+    if local_dataset:
+        dataset = load_from_disk(
+            dataset_name,
+        )
+    else:
+        dataset = load_dataset(
+            dataset_name,
+            split=split,
+            num_proc=num_workers if not streaming else None,
+            streaming=streaming,
+        )
+    if streaming:
+        logging.info("Loading the dataset in streaming mode")
+        if not (shuffle_buffer and size_valid_set):
+            raise ValueError("size_valid_set must be set when streaming is enabled")
+        valid_data = dataset.take(size_valid_set)
+        train_data = dataset.skip(size_valid_set)
+        train_data = train_data.shuffle(buffer_size=shuffle_buffer, seed=None)
+    else:
+        if not validation_split:
+            raise ValueError("validation_split must be set when streaming is disabled")
+        dataset = dataset.train_test_split(test_size=validation_split, seed=None)
+        train_data = dataset["train"]
+        valid_data = dataset["test"]
+        logging.info(f"Size of the train set: {len(train_data)}. Size of the validation set: {len(valid_data)}")
+
+    return train_data, valid_data
+
+
+def chars_token_ratio(
+    dataset: Dataset,
+    tokenizer: PreTrainedTokenizerBase,
+    formatting_func: Callable[[Dict[str, Any]], str],
+    sample_size: int = 400,
+) -> float:
+    """
+    Estimate the average number of characters per token in the dataset.
+    """
+    total_characters, total_tokens = 0, 0
+    for _, example in tqdm(zip(range(sample_size), iter(dataset), strict=False), total=sample_size):
+        text = formatting_func(example)
+        total_characters += len(text)
+        if tokenizer.is_fast:
+            total_tokens += len(tokenizer(text).tokens())
+        else:
+            total_tokens += len(tokenizer.tokenize(text))
+
+    return total_characters / total_tokens
+
+
+def parse_args() -> argparse.Namespace:
+    parser = argparse.ArgumentParser()
+    parser.add_argument("--model_name", type=str, default="HuggingFaceH4/zephyr-7b-beta", help="the model name")
+    parser.add_argument("--log_with", type=str, help="tracker name (wandb, mlflow, ..etc)")
+    parser.add_argument(
+        "--dataset_name",
+        type=str,
+        required=True,
+        help=(
+            "The dataset name corresponding to one sitting on huggingface or a local one"
+            "If local, be sure to set the local_dataset flag"
+        ),
+    )
+    parser.add_argument("--local_dataset", action="store_true", help="whether to use a local dataset")
```

Comment:
Just curious: is this the normal way of doing things in HF?  

I thought the normal hugging face approach is to "just figure it out".  Eg, try to load a local dataset, then fallback to loading from HF hub (or vice-versa).